### PR TITLE
SE: Implicit Null constraint for VB.NET declarations and ReDim support

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Related to #5973

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 1]
```diff
@@ -148,7 +148,7 @@ private ExplodedNode ProcessBranch(ExplodedNode node, ControlFlowBranch branch)
 
         private ProgramState ProcessBranchState(ControlFlowBranch branch, ProgramState state)
         {
-            foreach (var local in branch.EnteringRegions.SelectMany(x => x.Locals))
+            foreach (var local in branch.EnteringRegions.SelectMany(x => x.Locals).Where(x => x.Language == LanguageNames.VisualBasic)) // This is needed as we don't support tuple deconstructions yet
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:151]: Is it necessary to iterate all the way through if we are not VB? Can a region span multiple languages? If not, we can bail out after the first local.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 2]
```diff
@@ -713,5 +713,23 @@ public void EventReference_SetsNotNull()
             validator.ValidateTag("AfterAdd", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
             validator.ValidateTag("AfterRemove", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
         }
+
+        [TestMethod]
+        public void ReDim_SetsNotNull()
+        {
+            const string code = @"
+Dim First(), Second(), Third(4242) As Object
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:721]: Add an array declaration with an initializer
```
Dim arr() = New Object() { }
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 3]
```diff
@@ -713,5 +713,23 @@ public void EventReference_SetsNotNull()
...
+ReDim First(42), Second(1042), Third(4444), Arg.FieldArray(42)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:722]: Add a sample with `Preserve`

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:721]: Add Tag calls, so we can see `First` and `Second` are null but `Third` isn't before the ReDim.

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:721]: Add a syntax variation
`Dim arr as Object()`

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs [hunk 4]
```diff
@@ -274,8 +274,16 @@ public void Execute_ClearsCapturesAfterBranching()
         }
 
         [TestMethod]
-        public void Execute_LocalScopeRegion_AssignDefaultBoolConstraint() =>
-            SETestContext.CreateVB(@"Dim B As Boolean : Tag(""B"", B)").Validator.ValidateTag("B", x => x.HasConstraint(BoolConstraint.False).Should().BeTrue());
+        public void Execute_LocalScopeRegion_Boolean_AssignDefaultBoolConstraint() =>
+            SETestContext.CreateVB(@"Dim Value As Boolean : Tag(""Value"", Value)").Validator.ValidateTag("Value", x => x.HasConstraint(BoolConstraint.False).Should().BeTrue());
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs:278]: Add a test with an initializer to `True`

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs [hunk 5]
```diff
@@ -274,8 +274,16 @@ public void Execute_ClearsCapturesAfterBranching()
...
+
+        [TestMethod]
+        public void Execute_LocalScopeRegion_ReferenceType_AssignDefaultNullConstraint() =>
+            SETestContext.CreateVB(@"Dim Value As Exception : Tag(""Value"", Value)").Validator.ValidateTag("Value", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs:282]: Add a test with an initializer

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs [hunk 6]
```diff
@@ -274,8 +274,16 @@ public void Execute_ClearsCapturesAfterBranching()
...
+
+        [TestMethod]
+        public void Execute_LocalScopeRegion_Struct_NoAction() =>
+            SETestContext.CreateVB(@"Dim Value As Integer : Tag(""Value"", Value)").Validator.ValidateTag("Value", x => x.Should().BeNull());   // Will get NotNull in the future
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs:286]: Add a test with an uninitialized `Integer()` and an initialized `Integer(1)`

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
I miss some test about initializers overriding the constraint set by the "DefaultType"

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:721]: `Dim arr() = New Object() { }` is not a ReDim operation but ArrayInitializer

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs:278]: That would not be default anymore. Assignment is a different story

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs (hunk 5 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs:282]: Assignment is tested separately, see `SimpleAssignment_ToLocalVariable_FromLiteral`

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.cs:286]: That is now (a bit indirectly) tested in `ReDim_SetsNotNull`

**Review by martin-strecker-sonarsource — APPROVED** [review]

**Pull request merged**